### PR TITLE
Add Squash 0.7.0 recipe

### PR DESCRIPTION
**Pull request by rolando**
Squash is a compression abstraction library that provides an unified
interface for several compression algorithms.

Patches are either fixes from upstream or address OSX/clang compilation errors fixes. This version is the latest release from Squash[1].

[1] https://quixdb.github.io/squash/

--

This depends on:

- [x] #2345 Ragel package

### CODE REVIEW

File: recipes/squash/build.sh [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+set -ex
+
+EXTRA_CMAKE_ARGS=""
+if [[ `uname` == 'Darwin' ]]; then
+  export LIBRARY_SEARCH_VAR=DYLD_FALLBACK_LIBRARY_PATH
+  # We override most toolchain variables to modify the min version for OSX.
+  export MACOSX_VERSION_MIN="10.7"
```
Comment (ocefpaf) [inline recipes/squash/build.sh]: I am not sure these are the recommended ones from the `toolchain`. Please double check to ensure consistency.

Comment (rolando) [inline recipes/squash/build.sh]: [toolchain](https://github.com/conda-forge/toolchain-feedstock/blob/master/recipe/activate.sh) uses min version `10.9`. Then `-stdlib=libc++`, `-headerpad_max_install_names' and `-lc++` are used in `toolchain` too.

I have reviewed other recipes and references like:
https://github.com/conda-forge/vigra-feedstock/blob/master/recipe/build.sh
https://github.com/conda-forge/protobuf-feedstock/blob/master/recipe/build.sh
https://github.com/conda-forge/flatbuffers-feedstock/blob/master/recipe/build.sh
https://ipscc.readthedocs.io/en/latest/build-sh-guidelines.html#clang-and-libc-dylib

And it makes sense to me to target older OSX version. Unless official conda-forge policy is to go with `toolchain` defaults, if that so I will remove this change.

**Comment by rolando** [comment]
Build fails because #2345 is merged but package is not yet released.

**Comment by ocefpaf** [comment]
> Build fails because #2345 is merged but package is not yet released.

You can keep an eye at https://travis-ci.org/conda-forge/staged-recipes/builds/199146925 and, once that finishes, re-start the PR here.

**Comment by rolando** [comment]
@ocefpaf The build https://travis-ci.org/conda-forge/staged-recipes/builds/199677810
create the repo https://github.com/conda-forge/ragel-feedstock
but it's empty.

It will be fixed automatically?

**Comment by ocefpaf** [comment]
> It will be fixed automatically?

It should. Travis-CI is awfully slow lately. Let's wait 1-2 days and then I will check it again.

**Comment by rolando** [comment]
Ragel feedstock has been published. New builds are on the way.

File: recipes/squash/build.sh (hunk 1 above)
Comment (jakirkham) [inline recipes/squash/build.sh]: Those are holdovers that need to be updated to use the `toolchain`. As is documented in issue ( https://github.com/conda-forge/toolchain-feedstock/issues/7 ), we actually can't effectively set the minimum version to 10.7. The best we can do is use 10.9 as it is the oldest SDK the XCode on Travis CI comes with.

Comment (rolando) [inline recipes/squash/build.sh]: @jakirkham Oh, I see. I have removed those variables from the build script.

File: recipes/squash/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    - patches/0007-Fix-negative-bit-left-shifting.patch
+
+build:
+  number: 0
+  skip: True  # [win]
+
+requirements:
+  build:
+    - toolchain
+    - gcc  # [linux]
```
Comment (jakirkham) [inline recipes/squash/meta.yaml]: So we supply `gcc` on Linux. Would recommend dropping this unless there is a special reason to keep it (e.g. needing Fortran or OpenMP).

**Comment by jakirkham** [comment]
Feel free to add `[skip appveyor]` to your commit messages if you are not building on Windows.

File: recipes/squash/meta.yaml (hunk 2 above)
Comment (rolando) [inline recipes/squash/meta.yaml]: IIRC, I was having issues with with the compiler provided in the ``linux-anvil`` image. Let me try sending a build without the `gcc` package.

**Comment by rolando** [comment]
@jakirkham Build is OK without `gcc` dependency for linux.

File: recipes/squash/build.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  EXTRA_CMAKE_ARGS="${EXTRA_CMAKE_ARGS} -DCMAKE_OSX_DEPLOYMENT_TARGET=${MACOSX_DEPLOYMENT_TARGET}"
+  EXTRA_CMAKE_ARGS="${EXTRA_CMAKE_ARGS} -DCMAKE_MACOSX_RPATH=ON"
+else
+  export LIBRARY_SEARCH_VAR=LD_LIBRARY_PATH
+fi
+
+# This flags come from project's travis build.
+export CFLAGS="${CFLAGS} -fno-omit-frame-pointer -fstack-protector-all"
+export CXXFLAGS="${CXXFLAGS} -fno-omit-frame-pointer -fstack-protector-all"
+eval export $LIBRARY_SEARCH_VAR=${PREFIX}/lib
```
Comment (jakirkham) [inline recipes/squash/build.sh]: This is also getting phased out. Setting the rpath explicitly (e.g. `-Wl,-rpath,${PREFIX}/lib`) is the recommended way to go and seems to work fine. The main driver behind this is the addition of SIP on 10.11. While we are not building on 10.11 yet, it is a matter of time before we do.

xref: https://github.com/conda-forge/conda-forge.github.io/issues/238

File: recipes/squash/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    - xz 5.2.*
+    - lzo 2.06
+    - snappy
+    - zlib 1.2.*
+
+  run:
+    - glib 2.51.*
+    - bzip2 1.0.*
+    - xz 5.2.*
+    - lzo 2.06
```
Comment (jakirkham) [inline recipes/squash/meta.yaml]: Do we need this tight of a pinning on `lzo`?

File: recipes/squash/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  url: https://github.com/quixdb/squash/releases/download/v{{ version }}/{{ name }}-{{ version }}.tar.bz2
+  {{ hash_type }}: {{ hash_val }}
+  patches:
+    - patches/0001-Remove-wrongly-placed-break-on-cmake-file.patch
+    - patches/0002-Add-directive-to-have-_SC_PAGE_SIZE-defined.patch
+    - patches/0003-Fix-squash-missing-endian-declarations-in-OSX.patch
+    - patches/0004-Fix-zconf.h-generation.patch
+    - patches/0005-Fix-lzham-compilation-on-OSX.patch
+    - patches/0006-Fix-negative-left-shifting.patch
+    - patches/0007-Fix-negative-bit-left-shifting.patch
```
Comment (jakirkham) [inline recipes/squash/meta.yaml]: Are these being or have they been sent upstream?

File: recipes/squash/meta.yaml (hunk 4 above)
Comment (rolando) [inline recipes/squash/meta.yaml]: No, I think just `lzo` would be ok.

File: recipes/squash/meta.yaml (hunk 5 above)
Comment (rolando) [inline recipes/squash/meta.yaml]: All of them are backported fixes, either from squash itself or upstream libraries (i.e.: from zlib, zpaq, etc).

File: recipes/squash/build.sh (hunk 3 above)
Comment (rolando) [inline recipes/squash/build.sh]: Good to know!

**Comment by rolando** [comment]
@jakirkham let me know if there is anything else you want to review.

**Comment by rolando** [comment]
@jakirkham The OSX build fails due to a bounds test in the BSC plugin (one out of two dozens of plugins). Linux tests pass OK.

Do you think can we allow that failure in OSX? (or skip the bounds tests entirely)

Locally also I have experienced tests failure with the `ms-compress` plugin in OSX. Given squash bundles those compression libraries, I think is out of the control of the main author whether some plugins fails under certain edge cases.  Related: https://github.com/quixdb/squash/issues/204

**Comment by jakirkham** [comment]
Is it possible issue ( https://github.com/quixdb/squash/issues/118 ) is related? In either event, we should probably raise the issue upstream to see if they have any ideas.

**Comment by jakirkham** [comment]
Could we patch out or otherwise selectively skip the failing test on OS X for now?

**Comment by rolando** [comment]
For the record, I've added OSX-only conditional skips for the following tests:

* bounds: bsc, density, lz4f, lz4g, lzham, lznt1, deflate, wflz, zpaq
* random-data: xpress-huffman
* threads: lznt1, xpress-huffman, compress

Worth nothing that some of the codecs (i.e.: density, lz4f, lz4g lzham failed on Travis (XCode 6.3) but not in my machine (XCode 7.3). However, the recipe includes a naive test of compress->decompress of a text file just to ensure the library works in the most basic case.

Once we have the feedstock created we can work on a recipe for the next 0.8 release or, depending on the users' demand, a 0.7.1 release (in the upstream repository).

@jakirkham Finally got the OSX build OK! 🎉 

CC: @nemequ

**Comment by jakirkham** [comment]
Thanks @rolando.

File: recipes/squash/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    - xz 5.2.*
+    - lzo  # [linux]
+    - snappy
+    - zlib 1.2.*
+
+  run:
+    - glib 2.51.*
+    - bzip2 1.0.*
+    - xz 5.2.*
+    - lzo  # [linux]
```
Comment (jakirkham) [inline recipes/squash/meta.yaml]: Could you please remind me why this is Linux only?

Comment (rolando) [inline recipes/squash/meta.yaml]: It wasn't at first, but then I noticed the only recipe that uses `lzo` is `pytables` and uses `[linux]` selector. Then noticed that the `osx-64` package is actually empty:

```
$ curl -L https://anaconda.org/anaconda/lzo/2.06/download/osx-64/lzo-2.06-0.tar.bz2 | tar jtvf -
-rw-r--r--  0 ilan   admin       0 Sep 11  2012 .index/lzo-2.06-0
-rw-r--r--  0 ilan   admin      18 Sep 11  2012 info/files
-rw-r--r--  0 ilan   admin       0 Sep 11  2012 info/has_prefix
-rw-r--r--  0 ilan   admin     142 Sep 11  2012 info/index.json
-rw-r--r--  0 ilan   admin     159 Sep 11  2012 info/meta
```

As far I can tell from the build logs, `pytables` builds without `lzo` support and `squash` uses the vendored `lzo` code. 

I have no idea why the `lzo` package is empty for `osx-64`. But stating that we only use it for `linux` seems a good thing to me.

Comment (jakirkham) [inline recipes/squash/meta.yaml]: That's not great. Guessing `lzo` comes from `defaults`. We should probably let them know.

Though it might be good for us to get our own `lzo` package as well.

xref: https://github.com/conda-forge/staged-recipes/issues/2428
